applications/tari_validator_node/src/p2p/services/mempool/service.rs

```diff
@@ -101,35 +116,99 @@ impl MempoolService {
         }
 
         {
-            let mut access = self.transactions.lock().unwrap();
+            let access = self.transactions.lock().unwrap();
             // TODO: O(n)
             if access.iter().any(|(tx, _)| tx.hash() == transaction.hash()) {
                 info!(
                     target: LOG_TARGET,
                     "🎱 Transaction {} already in mempool",
                     transaction.hash()
                 );
-                return;
+                return Err(MempoolError::TransactionAlreadyExists);
             }
+        }
 
-            access.push((transaction.clone(), None));
+        {
+            let current_node_pubkey = self.node_identity.public_key().clone();
+            let mut should_process_txn = false;
+
+            for sid in &shards {
+                if self
+                    .epoch_manager
+                    .is_validator_in_committee_for_current_epoch(*sid, current_node_pubkey.clone())
+                    .await
+                    .map_err(|e| MempoolError::EpochManagerError(Box::new(e)))?
+                {
+                    should_process_txn = true;
+                    break;
+                }
+            }
+
+            let mut access = self.transactions.lock().unwrap();
+
+            if should_process_txn {
+                access.push((transaction.clone(), None));
+            } else {
```

Comment:
We still want to broadcast the transaction even if it's not for this node

Reply:
@sdbondi, only for the VN committee, is that right ?

Reply:
Currently, in L164

```rs
 match self.propagate_transaction(&transaction, &shards).await {
            Ok(()) => (),
            Err(e) => error!(
                target: LOG_TARGET,
                "Unable to propagate transaction among peers: {}",
                e.to_string()
            ),
        }
        ```
        
        we broadcast the transaction to peer VNs in the same committee as the current VN.

Reply:
Yup, that's right. 

We have three steps
1. validate and store the transaction if we're involved
2. propagate/broadcast the valid transaction to the shards involved (whether we are involved or not)
3. Send the transactions that need to be processed to the hotstuff worker (line 173)

For number 3, I think filter out to only the shards this node is involved with so that it only sends those to the worker (there could be 0 shards involved  which is still correct)